src/workerd/api/kv-instrumentation-test.js

```diff
@@ -20,28 +21,58 @@ export default {
       })
     );
 
+    // Store the top-level span ID from the onset event
+    if (event.event.type === 'onset' && event.event.spanId) {
+      topLevelSpanIds.set(event.invocationId, event.event.spanId);
+    }
+
     // Accumulate the span info for easier testing
     return (event) => {
-      // span ids are simple counters for tests, but invocation ID allows us to differentiate them
-      let spanKey = event.invocationId + event.spanContext.spanId;
+      // For spanOpen events, the new span ID is in event.event.spanId
+      // For other events, they reference an existing span via event.spanContext.spanId
+      let spanKey;
       switch (event.event.type) {
         case 'spanOpen':
-          spans.set(event.invocationId + event.event.spanId, {
+          // spanOpen creates a new span with ID in event.event.spanId
+          spanKey = event.invocationId + event.event.spanId;
+          spans.set(spanKey, {
             name: event.event.name,
           });
           break;
         case 'attributes': {
+          // Filter out top-level attributes events (jsRpcSession span)
+          const topLevelSpanId = topLevelSpanIds.get(event.invocationId);
+          if (topLevelSpanId && event.spanContext.spanId === topLevelSpanId) {
+            // Ignore attributes for the top-level span
+            break;
+          }
+
+          // attributes references an existing span via spanContext.spanId
+          spanKey = event.invocationId + event.spanContext.spanId;
           let span = spans.get(spanKey);
+          if (!span) {
+            throw new Error(`Attributes event for unknown span: ${spanKey}`);
+          }
           for (let { name, value } of event.event.info) {
             span[name] = value;
           }
-          spans.set(spanKey, span);
           break;
         }
         case 'spanClose': {
+          // Filter out top-level spanClose events (jsRpcSession span)
+          const topLevelSpanId = topLevelSpanIds.get(event.invocationId);
+          if (topLevelSpanId && event.spanContext.spanId === topLevelSpanId) {
```

Comment:
Note that we only ever need to filter out attributes events, not spanClose events (the Onset is closed by Outcome, it is not closed by anything else), so this branch is not needed.
spanKey should also stay as-is – currently we're duplicating how it is being computed, which makes the code less legible.